urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/json/rpc/JSONRPCHandler.java

```diff
@@ -0,0 +1,802 @@
+	reader.close();
+
+	return response;
+    }
+
+    /**
+     * Calls the method as specified in the JSON-RPC request from the reader.
+     *
+     * @param reader A reader to read the JSON-RPC request from.
+     * @return The JSON response (may be an error response).
+     * @throws NullPointerException If the reader is null.
+     **/
+    public JSONRPCResponse call( Reader reader ) 
+	throws NullPointerException {
+
+	if( reader == null )
+	    throw new NullPointerException( "Cannot handle RPC request from null reader." );
+
+
+	try {
+
+	    //Reader reader = new StringReader( jsonString );
+	    JSONRPCRequest request = this.buildRPCRequest( reader );
+	    //reader.close();	    
+	    return this.execute( request );
+
+	} catch( IOException e ) {
+	    //throw new JSONRPCException( "Cannot read from json string: " + e.getMessage() );
+	    return createErrorResponse( e, 
+					"IO error: " + e.getMessage(), 
+					JSONRPCError.CODE_SERVER_ERROR_MIN,
+					null   // no ID
+					);
+	} catch( JSONSyntaxException e ) {
+	    return createErrorResponse( e, 
+					"Parse error: " + e.getMessage(), 
+					JSONRPCError.CODE_PARSE_ERROR,
+					null   // no ID
+					);
+	} catch( JSONException e ) {
+	    return createErrorResponse( e, 
+					"Invalid request: " + e.getMessage(), 
+					JSONRPCError.CODE_INVALID_REQUEST,
+					null   // no ID
+					);
+	} catch( JSONRPCException e ) {
+	    return createErrorResponse( e, 
+					"JSON-RPC error: " + e.getMessage(), 
+					JSONRPCError.CODE_SERVER_ERROR_MIN,
+					null   // no ID
+					);
+	}
+
+    }
+    
+    
+    /**
+     * Calls the method as specified in the JSON-RPC request.
+     *
+     * @param request The request to execute.
+     * @return The JSON response (may be an error response).
+     * @throws NullPointerException If the request is null.
+     **/
+    public JSONRPCResponse call( JSONRPCRequest request ) 
+	throws NullPointerException {
+
+	if( request == null )
+	    throw new NullPointerException( "Cannot execute null-requests." );
+
+
+
+	try {
+	    
+	    return this.execute( request );
+
+	} catch( JSONRPCException e ) {
+
+	    return createErrorResponse( e, 
+					"Unexpected JSON-RPC error: " + e.getMessage(), 
+					JSONRPCError.CODE_SERVER_ERROR_MIN,
+					request.getID()
+					);
+
+	} catch( JSONException e ) {
+	    
+	    return createErrorResponse( e, 
+					"Unexpected JSON error: " + e.getMessage(), 
+					JSONRPCError.CODE_SERVER_ERROR_MIN,
+					request.getID()
+					);
+
+	}
+    }
+    
+
+    /**
+     * This is the actual method executing JSON-RPC requests.
+     *
+     * It may still throw some exceptions.
+     *
+     * @param request The request to execute.
+     * @return A JSONRPCResponse - if no fundamental exceptions occured.
+     * @param JSONException    Indicates unexpected JSON errors.
+     * @param JSONRPCException Indicates unexpected JSON RPC errors.
+     **/
+    private JSONRPCResponse execute( JSONRPCRequest request )
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.